File: presentation/src/test/java/com/example/presentation/ui/viewmodel/MainViewModelTest.java
package com.example.presentation.ui.viewmodel;

import androidx.arch.core.executor.testing.InstantTaskExecutorRule;

import com.example.domain.exception.CommonException;
import com.example.domain.exception.NetworkAccessException;
import com.example.domain.interactor.IAstronomyPictureInteractor;
import com.example.domain.model.APODEntity;
import com.example.presentation.utils.DateUtils;
import com.example.presentation.utils.scheduler.IBaseSchedulerProvider;
import com.example.presentation.utils.scheduler.TrampolineSchedulerProvider;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;

import io.reactivex.Single;

import static org.junit.Assert.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyZeroInteractions;
import static org.mockito.Mockito.when;

public class MainViewModelTest {

    @Rule
    public InstantTaskExecutorRule mInstantTaskExecutorRule = new InstantTaskExecutorRule();

    private APODEntity mEnteredEntityPicture;
    private APODEntity mEnteredEntityVideo;
    private final int mCurrentPositionPageAdapter = 0;
    private final String testDate = DateUtils.getDateOffset(mCurrentPositionPageAdapter);


    private IAstronomyPictureInteractor mInteractor;
    private MainViewModel mViewModel;
    private IBaseSchedulerProvider mSchedulerProvider;

    @Before
    public void setUp() {

        mEnteredEntityPicture = new APODEntity(
                testDate,
                "some explanation",
                "title",
                "https://apod.nasa.gov/apod/image/picture.png",
                "copyright"
        );

        mEnteredEntityVideo = new APODEntity(
                testDate,
                "some explanation",
                "title",
                "https://www.youtube.com/embed/ofZTOxC9JQ4?rel=0",
                "copyright"
        );

        mInteractor = mock(IAstronomyPictureInteractor.class);
        mSchedulerProvider = new TrampolineSchedulerProvider();
        mViewModel = new MainViewModel(mInteractor, mSchedulerProvider, mCurrentPositionPageAdapter);

    }

    @Test
    public void testShowInformationPicture() {
        when(mInteractor.getAstronomyPicture(any())).thenReturn(Single.fromCallable(() -> mEnteredEntityPicture));

        mViewModel.showInformation();

        verify(mInteractor).getAstronomyPicture(testDate);
        verifyZeroInteractions(mInteractor);
        assertEquals(true, mViewModel.isLoadingPicture().getValue());
        assertEquals(false, mViewModel.isLoadingData().getValue());
        assertFalse(mViewModel.isErrorVisible().get());
        assertEquals(false, mViewModel.isNetworkError().getValue());
        assertEquals(true, mViewModel.isPictureView().getValue());
        assertEquals(mEnteredEntityPicture.getTitle(), mViewModel.getTitle().get());
        assertEquals(mEnteredEntityPicture.getExplanation(), mViewModel.getExplanation().get());
        assertEquals(mEnteredEntityPicture.getUrl(), mViewModel.getUrlPicture().get());
        assertEquals(mEnteredEntityPicture.getCopyright(), mViewModel.getCopyright().get());
    }

    @Test
    public void testShowInformationVideo() {
        when(mInteractor.getAstronomyPicture(any())).thenReturn(Single.fromCallable(() -> mEnteredEntityVideo));

        mViewModel.showInformation();

        verify(mInteractor).getAstronomyPicture(testDate);
        verifyZeroInteractions(mInteractor);
        assertEquals(true, mViewModel.isLoadingPicture().getValue());
        assertEquals(false, mViewModel.isLoadingData().getValue());
        assertFalse(mViewModel.isErrorVisible().get());
        assertEquals(false, mViewModel.isNetworkError().getValue());
        assertEquals(false, mViewModel.isPictureView().getValue());
        assertEquals(mEnteredEntityVideo.getTitle(), mViewModel.getTitle().get());
        assertEquals(mEnteredEntityVideo.getExplanation(), mViewModel.getExplanation().get());
        assertEquals(mEnteredEntityVideo.getCopyright(), mViewModel.getCopyright().get());
    }

    @Test
    public void testShowInformationNetworkAccessException() {
        when(mInteractor.getAstronomyPicture(testDate)).thenReturn(Single.error(new NetworkAccessException(new Throwable())));

        mViewModel.showInformation();

        verify(mInteractor).getAstronomyPicture(testDate);
        verifyZeroInteractions(mInteractor);
        assertEquals(true, mViewModel.isLoadingPicture().getValue());
        assertEquals(false, mViewModel.isLoadingData().getValue());
        assertTrue(mViewModel.isErrorVisible().get());
        assertEquals(true, mViewModel.isNetworkError().getValue());
        assertNull(mViewModel.getTitle().get());
        assertNull(mViewModel.getExplanation().get());
        assertNull(mViewModel.getUrlPicture().get());
        assertNull(mViewModel.getCopyright().get());

    }

    @Test
    public void testShowInformationCommonException() {
        when(mInteractor.getAstronomyPicture(testDate)).thenReturn(Single.error(new CommonException(new Throwable())));

        mViewModel.showInformation();

        verify(mInteractor).getAstronomyPicture(testDate);
        verifyZeroInteractions(mInteractor);
        assertEquals(true, mViewModel.isLoadingPicture().getValue());
        assertEquals(false, mViewModel.isLoadingData().getValue());
        assertTrue(mViewModel.isErrorVisible().get());
        assertEquals(false, mViewModel.isNetworkError().getValue());
        assertNull(mViewModel.getTitle().get());
        assertNull(mViewModel.getExplanation().get());
        assertNull(mViewModel.getUrlPicture().get());
        assertNull(mViewModel.getCopyright().get());
    }


}

File: presentation/src/test/java/com/example/presentation/ui/viewmodel/MainViewModelTest.kt
package com.example.presentation.ui.viewmodel

import androidx.arch.core.executor.testing.InstantTaskExecutorRule
import com.example.domain.exception.CommonException
import com.example.domain.exception.NetworkAccessException
import com.example.domain.interactor.IAstronomyPictureInteractor
import com.example.domain.model.APODEntity
import com.example.presentation.ui.viewmodel.utils.MainCoroutineScopeRule
import com.example.presentation.utils.DateUtils.getDateOffset
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.runBlockingTest
import org.junit.Assert
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.mockito.Mockito

@ExperimentalCoroutinesApi
class MainViewModelTest {
    @get:Rule
    val coroutineScope = MainCoroutineScopeRule()
    @get:Rule
    val instantTaskExecutorRule = InstantTaskExecutorRule()

    private lateinit var enteredEntityPicture: APODEntity
    private lateinit var enteredEntityVideo: APODEntity
    private val currentPositionPageAdapter = 0
    private val testDate = getDateOffset(currentPositionPageAdapter)
    private lateinit var interactor: IAstronomyPictureInteractor
    private lateinit var viewModel: MainViewModel
    @Before
    fun setUp() {
        enteredEntityPicture = APODEntity(
                testDate,
                "some explanation",
                "title",
                "https://apod.nasa.gov/apod/image/picture.png",
                "copyright"
        )
        enteredEntityVideo = APODEntity(
                testDate,
                "some explanation",
                "title",
                "https://www.youtube.com/embed/ofZTOxC9JQ4?rel=0",
                "copyright"
        )
        interactor = Mockito.mock(IAstronomyPictureInteractor::class.java)
        viewModel = MainViewModel(interactor, currentPositionPageAdapter)
    }

    @Test
    fun testShowInformationPicture() {
        runBlockingTest {
            Mockito.`when`(interactor.getAstronomyPicture(testDate)).thenReturn(enteredEntityPicture)
            viewModel.showInformation()

            Mockito.verify(interactor).getAstronomyPicture(testDate)

            Mockito.verifyZeroInteractions(interactor)

            Assert.assertEquals(true, viewModel.isLoadingPicture.value)
            Assert.assertEquals(false, viewModel.isLoadingData.value)
            Assert.assertFalse(viewModel.isErrorVisible.get())
            Assert.assertEquals(false, viewModel.isNetworkError.value)
            Assert.assertEquals(true, viewModel.isPictureView.value)
            Assert.assertEquals(enteredEntityPicture.title, viewModel.title.get())
            Assert.assertEquals(enteredEntityPicture.explanation, viewModel.explanation.get())
            Assert.assertEquals(enteredEntityPicture.url, viewModel.urlPicture.get())
            Assert.assertEquals(enteredEntityPicture.copyright, viewModel.copyright.get())
        }
    }

    @Test
    fun testShowInformationVideo() {
        runBlockingTest {
            Mockito.`when`(interactor.getAstronomyPicture(testDate)).thenReturn(enteredEntityVideo)
            viewModel.showInformation()

            Mockito.verify(interactor).getAstronomyPicture(testDate)

            Mockito.verifyZeroInteractions(interactor)

            Assert.assertEquals(true, viewModel.isLoadingPicture.value)
            Assert.assertEquals(false, viewModel.isLoadingData.value)
            Assert.assertFalse(viewModel.isErrorVisible.get())
            Assert.assertEquals(false, viewModel.isNetworkError.value)
            Assert.assertEquals(false, viewModel.isPictureView.value)
            Assert.assertEquals(enteredEntityVideo.title, viewModel.title.get())
            Assert.assertEquals(enteredEntityVideo.explanation, viewModel.explanation.get())
            Assert.assertEquals(enteredEntityVideo.copyright, viewModel.copyright.get())
        }
    }

    @Test
    fun testShowInformationNetworkAccessException() {
        runBlockingTest {
            Mockito.`when`(interactor.getAstronomyPicture(testDate)).thenAnswer { throw NetworkAccessException(Throwable()) }
            viewModel.showInformation()

            Mockito.verify(interactor).getAstronomyPicture(testDate)

            Mockito.verifyZeroInteractions(interactor)

            Assert.assertEquals(true, viewModel.isLoadingPicture.value)
            Assert.assertEquals(false, viewModel.isLoadingData.value)
            Assert.assertTrue(viewModel.isErrorVisible.get())
            Assert.assertEquals(true, viewModel.isNetworkError.value)
            Assert.assertNull(viewModel.title.get())
            Assert.assertNull(viewModel.explanation.get())
            Assert.assertNull(viewModel.urlPicture.get())
            Assert.assertNull(viewModel.copyright.get())
        }
    }

    @Test
    fun testShowInformationCommonException() {
        runBlockingTest {
            Mockito.`when`(interactor.getAstronomyPicture(testDate)).thenAnswer { CommonException(Throwable()) }
            viewModel.showInformation()

            Mockito.verify(interactor).getAstronomyPicture(testDate)

            Mockito.verifyZeroInteractions(interactor)

            Assert.assertEquals(true, viewModel.isLoadingPicture.value)
            Assert.assertEquals(false, viewModel.isLoadingData.value)
            Assert.assertTrue(viewModel.isErrorVisible.get())
            Assert.assertEquals(false, viewModel.isNetworkError.value)
            Assert.assertNull(viewModel.title.get())
            Assert.assertNull(viewModel.explanation.get())
            Assert.assertNull(viewModel.urlPicture.get())
            Assert.assertNull(viewModel.copyright.get())
        }
    }
}